AVX512 and VPCLMULQDQ based CRC-32 and CRC-32C

dr-m opened the pull request:
*Issue #, if available:* #72

*Description of changes:* This implementation in `crc32_avx512.c` is based on `crc32_refl_by16_vclmul_avx512` in https://github.com/intel/intel-ipsec-mb/ with some optimizations.

Some of the code is based on #72. Because this repository appears to depend on definitions in other repositories, I was not able to compile and test this locally. I merely tested that `gcc -O2 -std=c11 -c crc32_avx512.c` produces reasonable looking output.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

pbadari commented:
1) Bunch of compile failures on my avx512 enabled system.
2) Instead of implementing aws_checksums_hw_detect(), we should use the one provided in aws-c-common:
eg. 
`
+    if (AWS_UNLIKELY(!detection_performed)) {
+        detected_avx512 = aws_cpu_has_feature(AWS_CPU_FEATURE_AVX512);
+        detected_vpclmulqdq = aws_cpu_has_feature(AWS_CPU_FEATURE_VPCLMULQDQ);
+        detection_performed = true;
`

dr-m commented:
> 1. Bunch of compile failures on my avx512 enabled system.
> 
>     2. Instead of implementing aws_checksums_hw_detect(), we should use the one provided in aws-c-common:
>        eg.
> 
> 
> ```
> +    if (AWS_UNLIKELY(!detection_performed)) {
> +        detected_avx512 = aws_cpu_has_feature(AWS_CPU_FEATURE_AVX512);
> +        detected_vpclmulqdq = aws_cpu_has_feature(AWS_CPU_FEATURE_VPCLMULQDQ);
> +        detection_performed = true;
> ```

Thank you, I was aware that this repository depends on something else, and now I know that it is https://github.com/awslabs/aws-c-common/. As far as I understand, `aws_cpu_has_feature(AWS_CPU_FEATURE_AVX512)` is not a sufficient condition, because it only refers to the AVX512F feature flag. The implementation depends on AVX512DQ, AVX512BW and AVX512VL as well.

dr-m commented:
> 1. I still see lot of warning when compiling

Sorry, I somehow overlooked those warnings. Fixed in c69772cb7034675059e5359b007112ed4d05acf9.

>     2. Currently the new code doesn't get executed and needs something like..

Right, we would need that for the zlib `crc32()` compatible function. The CRC-32C acceleration should work without any changes, because anything that supports AVX512F would also support SSE4.2.

I will try to refactor the ISA extension detection. I think that `s_crc32_fn_ptr` and `s_crc32c_fn_ptr` should point to one specific implementation, so there will be no single entry point `aws_checksums_crc32_hw` or `aws_checksums_crc32c_hw` but specific implementations, such as `aws_checksums_crc32_sse42` or `aws_checksums_crc32_avx512`. The ISA extension detection functions would return a pointer to a specific implementation.

dr-m commented:
While refactoring the ISA extension detection in 7c6a87a0b282e40a6f62970b3e7a962a05cc7896, I found a bug in your existing implementation, in case the carry-less multiplication is not available:
```diff
diff --git a/source/intel/crc_hw.c b/source/intel/crc_hw.c
--- a/source/intel/crc_hw.c
+++ b/source/intel/crc_hw.c
@@ -95,13 +95,13 @@ uint32_t aws_checksums_crc32c_hw(const uint8_t *input, int length, uint32_t prev
         crc = (uint32_t)_mm_crc32_u8(crc, *input++);
     }
 
-    if (detected_sse42 && detected_clmul) {
+    if (0) {
         return aws_checksums_crc32c_sse42(input, length, crc);
     }
 
     /* Spin through remaining (aligned) 8-byte chunks using the CRC32Q quad word instruction */
     while (length >= (int)sizeof(slice_ptr_int_type)) {
-        crc = (uint32_t)crc_intrin_fn(crc, *input);
+        crc = (uint32_t)crc_intrin_fn(crc, *(const slice_ptr_int_type*)input);
         input += sizeof(slice_ptr_int_type);
         length -= (int)sizeof(slice_ptr_int_type);
     }
```
The `crc32` instructions that this function is using are already kind of part of SSE 4.2; they were introduced in the Intel Nehalem microarchitecture. So, the check `detected_sse42` is already kind of late here. With the above fix, also the generated code looks saner: `crc32q` is being invoked on a memory argument, instead of on a register that was `movzbl`’d from memory.

Note: my CPU detection depends on the following patch to awslabs/aws-c-common@6ebf3bc45ffcaa3709a8bab8dc373d82e7cdd4cc:
```diff
diff --git a/source/arch/intel/cpuid.c b/source/arch/intel/cpuid.c
index 465fccd1..03c52993 100644
--- a/source/arch/intel/cpuid.c
+++ b/source/arch/intel/cpuid.c
@@ -88,9 +88,10 @@ static bool s_has_avx2(void) {
 static bool s_has_avx512(void) {
     uint32_t abcd[4];
 
-    /* Check AVX512F:
-     * CPUID.(EAX=07H, ECX=0H):EBX.AVX512[bit 16]==1 */
-    uint32_t avx512_mask = (1 << 16);
+    /* Check AVX512 flags.
+     * CPUID.(EAX=07H, ECX=0H):EBX */
+    uint32_t avx512_mask = 1 << 16 /* AVX512F */ | 1 << 17 /* AVX512DQ */ |
+      1U << 30 /* AVX512BW */ | 1U << 31 /* AVX512VL */;
     aws_run_cpuid(7, 0, abcd);
     if ((abcd[1] & avx512_mask) != avx512_mask) {
         return false;
```
It could be clearer to rename `AWS_CPU_FEATURE_AVX512` to something like `AWS_CPU_FEATURE_AVX10_1_512` and make it cover at least these 4 AVX512 flags. To my understanding, `avx10.1-512` is just a new name for a bunch of existing AVX512 ISA extensions, and `avx10.1-256` is a name for a 256-bit subset. Any further extensions `avx10.2-256` and possibly `avx10.2-512` would introduce new `cpuid` flags.

By the way, I see that `source/crc.c` is being compiled without any ISA extension flags. But `source/intel/crc_hw.c` and `source/intel/asm/crc32c_sse42_asm.c` are in danger: you are enabling among others AVX2, AVX512F (but not other AVX512 flags) while compiling them. As far as I can tell, it would only be safe to enable `-msse4.2` for the first file, and `-msse4.2 -mpclmul` for the second one.

pbadari commented:
One more change needed:

```
--- a/source/crc.c
+++ b/source/crc.c
@@ -43,7 +43,7 @@ uint32_t aws_checksums_crc32c(const uint8_t *input, int length, uint32_t previou
 # ifdef AWS_HAVE_AVX512_INTRINSICS
         if (aws_cpu_has_feature(AWS_CPU_FEATURE_AVX512) &&
             aws_cpu_has_feature(AWS_CPU_FEATURE_VPCLMULQDQ))
-            s_crc32_fn_ptr = aws_checksums_crc32c_avx512;
+            s_crc32c_fn_ptr = aws_checksums_crc32c_avx512;
        else
 # endif
        if (aws_cpu_has_feature(AWS_CPU_FEATURE_SSE_4_2)) {

```

dr-m commented:
> One more change needed:

Thank you. My bad, I do not currently have access to AVX512 capable hardware. Fixed in b06f512648a769ce4613fd6a68b2fb464cd9fcea. I also fixed the ISA extension trouble in 5095452fd8e1e97626220402a046b94f47833187. I do not yet have a solution for skipping the compilation of the AVX512 code on older compilers (clang 7 or earlier, GCC 10 or older).  The simplest solution that I can come up with would be to merge the AVX512 code to `source/intel/crc_hw.c` and then use `#if` magic to enable the AVX512 only if the compiler is recent enough. (I’d bravely assume that there are no other compilers than GCC, MSVC, and clang and its derivatives.) That is what I ended up doing in MariaDB/server@9ec7819c585d139c8fe64d0f7f0f0f51dcafa01f.

The inaccurate `AWS_CPU_FEATURE_AVX512` still needs to be addressed.

dr-m commented:
> I do not yet have a solution for skipping the compilation of the AVX512 code on older compilers (clang 7 or earlier, GCC 10 or older).

Well, I guess that this is actually covered by the `AWS_HAVE_AVX512_INTRINSICS`  check in the top-level `CMakeCache.txt` as well as the corresponding `#ifdef` in the code. I would love to see some CI results.

> The inaccurate `AWS_CPU_FEATURE_AVX512` still needs to be addressed.

I would like to hear your opinion on this. We can also ignore this and assume that the combination of checking for AVX512F and VPCLMULQDQ is adequate, that is, all processors that implement those will also implement AVX512DQ, AVX512BW, and AVX512VL.

dr-m commented:
If https://en.wikipedia.org/wiki/AVX-512#CPUs_with_AVX-512 can be trusted, `VPCLMULQDQ` implies not only `AVX512F` but also `AVX512VL`, `AVX512BW`, `AVX512DQ`. Hence, testing for `AWS_CPU_FEATURE_AVX512` and `AWS_CPU_FEATURE_VPCLMULQDQ` could be acceptable in practice.

I squashed all commits to 2e84a92e0d9e9fffbf098310a258457d45d4c8d0.

dr-m commented:
I think that you should test your AVX512 detection logic by starting the Linux kernel with the `noxsave` option, so that any attempt to use AVX registers would result in SIGILL. I implemented this in dr-m/crc32_simd@e9e8c240280c67c595b9e6d60b513eb2cf2e88e9 and revised some incorrect logic in MariaDB/server#3471.

waahm7 commented:
Hi @dr-m,

Thank you for creating the PR and sorry for the delayed response. Unfortunately, we don't have the bandwidth to review this PR in this quarter. We have added it to our backlog and will let you know as soon as we have any updates.
